### PR TITLE
Add ESAPI tests

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
Ok, these are quite a lot of commits, but they are all so similar, that I did not see an advantage in splitting this into several PRs.

First we have a bunch of integration tests, with one bugfix that they reveiled.

Then we have some unit-tests.

### CODE REVIEW

File: test/integration/esys-policy-authorize.int.c [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                           ESYS_TR_NONE, ESYS_TR_NONE,
+                           &inSensitivePrimary, &inPublic,
+                           &outsideInfo, &creationPCR, &primaryHandle_handle,
+                           &outPublic, &creationData, &creationHash,
+                           &creationTicket);
+    goto_if_error(r, "Error esys create primary", error);
+
+    /*
+     * 2. Create a trial policy with PolicyAuthorized. The name primary key
+     *    will be passed and the primry key will be used to sign policies.
```
Comment (dantpm) [inline test/integration/esys-policy-authorize.int.c]: s/primry/primary/

File: test/integration/esys-policy-ticket.int.c [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &outPublic,
+                        &nameKeySign,
+                        &keyQualifiedName);
+    goto_if_error(r, "Error: ReadPublic", error);
+
+    /*
+     * 2. The policy expiration in 3600 seconds will be siged.
```
Comment (dantpm) [inline test/integration/esys-policy-ticket.int.c]: s/siged/signed/

File: test/integration/esys-verify-signature.int.c [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+#include "tss2_esys.h"
+#include "tss2_mu.h"
+
+#include "esys_iutil.h"
+#define LOGMODULE test
+#include "util/log.h"
+
+/*
+ * This test is intended to test the ESAPI signing and signature verifcation.
```
Comment (dantpm) [inline test/integration/esys-verify-signature.int.c]: s/verifcation/verification/

File: test/integration/esys-duplicate.int.c [hunk 4]
```diff
@@ -33,19 +33,20 @@
...
  */
 
 int
 test_invoke_esapi(ESYS_CONTEXT * esys_context)
 {
     uint32_t r = 0;
 
-  /*
+    /*
      * Firth the policy value to be able to use Esys_Duplicate for an object has to be
```
Comment (dantpm) [inline test/integration/esys-duplicate.int.c]: s/Firth/First/

File: test/integration/esys-encrypt-decrypt.int.c [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+#include "tss2_esys.h"
+
+#include "esys_iutil.h"
+#define LOGMODULE test
+#include "util/log.h"
+
+/*
+ * This test is intended to test the ESAPI function Esys_EncryptDecrypt.
+ * First a primary key is generated. This key will be uses as parent fo a
+ * symmetric key, wich will be used to encrypt and decrypt a tpm2b. The
```
Comment (dantpm) [inline test/integration/esys-encrypt-decrypt.int.c]: s/witch/which/

File: test/integration/esys-policy-nv-undefine-special.int.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+ * A policy has to be defined for the command UndefineSpaceSpecial.
+ * The special handling whether the auth value is not used in the HMAC
+ * response verification will be checked.
+ */
+int
+test_invoke_esapi(ESYS_CONTEXT * esys_context)
+{
+    uint32_t r = 0;
+    /*
+     * Firth the policy value for NV_UndefineSpaceSpecial has to be
```
Comment (dantpm) [inline test/integration/esys-policy-nv-undefine-special.int.c]: s/Firth/First/

File: test/integration/esys-policy-ticket.int.c [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        &signed_digest,
+        &inScheme,
+        &hash_validation,
+        &signature);
+    goto_if_error(r, "Error: Sign", error);
+
+    /*
+     * 3. A policy session will be created. Based on the signed policy the
+     *    ticket policySignedTicket will be created.
+     *    With this ticke the function Esys_PolicyTicket will be tested.
```
Comment (dantpm) [inline test/integration/esys-policy-ticket.int.c]: s/ticke/ticket/

**Comment by AndreasFuchsSIT** [comment]
Ok, all of typos found by @dantpm are fixed.
@tstruk how's looking to you ?

File: test/integration/esys-audit.int.c [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+    r = Esys_CreatePrimary(esys_context, ESYS_TR_RH_OWNER, ESYS_TR_PASSWORD,
+                           ESYS_TR_NONE, ESYS_TR_NONE, &inSensitivePrimary,
+                           &inPublic, &outsideInfo, &creationPCR,
+                           &signHandle, &outPublic, &creationData,
+                           &creationHash, &creationTicket);
+    goto_if_error(r, "Error esys create primary", error);
+
+    /* Start a audit session */
+    TPMA_SESSION sessionAttributes;
```
Comment (tstruk) [inline test/integration/esys-audit.int.c]: TPMA_SESSION sessionAttributes = 0;
instead of memset()?

File: test/integration/esys-audit.int.c [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    memset(&sessionAttributes, 0, sizeof sessionAttributes);
+    sessionAttributes |= TPMA_SESSION_CONTINUESESSION;
+    sessionAttributes |= TPMA_SESSION_AUDIT;
```
Comment (tstruk) [inline test/integration/esys-audit.int.c]: Or even better:

TPMA_SESSION sessionAttributes = TPMA_SESSION_CONTINUESESSION | TPMA_SESSION_AUDIT;

File: test/integration/esys-certify-creation.int.c [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        &qualifyingData,
+        creationHash,
+        &inScheme,
+        creationTicket,
+        &certifyInfo,
+        &signature);
+    goto_if_error(r, "Error: CertifyCreation", error);
+
+    r = Esys_FlushContext(esys_context,signHandle);
+    goto_if_error(r, "Error: FlushContext", error);
```
Comment (tstruk) [inline test/integration/esys-certify-creation.int.c]: return r; instead of goto?

File: test/integration/esys-certify.int.c [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        ESYS_TR_NONE,
+        &qualifyingData,
+        &inScheme,
+        &certifyInfo,
+        &signature
+        );
+    goto_if_error(r, "Error: Certify", error);
+
+    r = Esys_FlushContext(esys_context,signHandle);
+    goto_if_error(r, "Error: FlushContext", error);
```
Comment (tstruk) [inline test/integration/esys-certify.int.c]: return r?

File: test/integration/esys-commit.int.c [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        &P1,
+        &s2,
+        &y2,
+        &K,
+        &L,
+        &E,
+        &counter
+        );
```
Comment (tstruk) [inline test/integration/esys-commit.int.c]: Pass multiple params per line here please.

File: test/integration/esys-lock.int.c [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+#include "esys_iutil.h"
+#define LOGMODULE test
+#include "util/log.h"
+
+/* Test the ESAPI functions related to TPM locks  */
+
+int
+test_invoke_esapi(ESYS_CONTEXT * esys_context)
+{
+
```
Comment (tstruk) [inline test/integration/esys-lock.int.c]: Empty line not needed here

File: test/integration/esys-pp-commands.int.c [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        auth_handle,
+        ESYS_TR_PASSWORD,
+        ESYS_TR_NONE,
+        ESYS_TR_NONE,
+        &setList,
+        &clearList);
+
+    if (r == (TPM2_RC_WARN  | TPM2_RC_PP)) {
+        LOG_INFO("Command TPM2_PP_Commands requires physical presence.");
+        r = 77; /* Skip */
```
Comment (tstruk) [inline test/integration/esys-pp-commands.int.c]: Will this always be the case here? If so maybe r = 0 will be better?

File: test/integration/esys-tpm-tests.int.c [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+#include "esys_iutil.h"
+#define LOGMODULE test
+#include "util/log.h"
+
+/* Test the ESAPI functions for TPM tests */
+int
+test_invoke_esapi(ESYS_CONTEXT * esys_context)
+{
+
```
Comment (tstruk) [inline test/integration/esys-tpm-tests.int.c]: Extra empty line

File: test/integration/esys-tpm-tests.int.c [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    TPM2B_MAX_BUFFER *outData;
+    TPM2_RC testResult;
+
+    esys_context->state = _ESYS_STATE_INIT;
+    r = Esys_GetTestResult(esys_context,
+                           ESYS_TR_NONE,
+                           ESYS_TR_NONE, ESYS_TR_NONE, &outData, &testResult);
+    goto_if_error(r, "Error GetTestResult did fail", error);
+
+
```
Comment (tstruk) [inline test/integration/esys-tpm-tests.int.c]: extra empty line

File: test/integration/esys-policy-regression.int.c [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+
+    LOGBLOB_DEBUG(&policyDigest->buffer[0], policyDigest->size,
+                  "POLICY DIGEST");
+
+    if (policyDigest->size != 20 ||
+        memcmp(&policyDigest->buffer[0], &expected_digest->buffer[0],
+               policyDigest->size)) {
+        LOG_ERROR("Error: Policy%s digest did not match expected policy.",
+                  comment);
+        return false;
```
Comment (tstruk) [inline test/integration/esys-policy-regression.int.c]: we have a memleak here.

File: test/integration/esys-tr-fromTpmPublic-nv.int.c [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    r = Esys_TR_Close(ectx, &nvHandle);
+    goto_if_error(r, "TR close on nv object", error);
+
+    r = Esys_TR_FromTPMPublic(ectx, TPM2_NV_INDEX_FIRST,
+                              ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+                              &nvHandle);
+    goto_if_error(r, "TR from TPM public", error);
+
+    r = Esys_TR_GetName(ectx, nvHandle, &name2);
+    goto_if_error(r, "TR get name", error);
```
Comment (tstruk) [inline test/integration/esys-tr-fromTpmPublic-nv.int.c]: memleak here for name1

File: test/integration/esys-tr-fromTpmPublic-nv.int.c [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+    r = Esys_NV_UndefineSpace(ectx, ESYS_TR_RH_OWNER, nvHandle,
+                              ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE);
+    goto_if_error(r, "NV UndefineSpace", error);
```
Comment (tstruk) [inline test/integration/esys-tr-fromTpmPublic-nv.int.c]: memleak

File: test/integration/esys-tr-fromTpmPublic-nv.int.c [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+    if (name1->size != name2->size ||
+        memcmp(&name1->name[0], &name2->name[0], name1->size) != 0)
+    {
+        LOG_ERROR("Names mismatch between NV_GetPublic and TR_GetName");
+        return 1;
```
Comment (tstruk) [inline test/integration/esys-tr-fromTpmPublic-nv.int.c]: memleak

File: test/integration/esys-tr-fromTpmPublic-key.int.c [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+    r = Esys_EvictControl(ectx, ESYS_TR_RH_OWNER, keyHandle,
+                          ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+                          TPM2_PERSISTENT_FIRST, &keyHandle);
+    goto_if_error(r, "EvictControl delete", error);
+
+    if (name1->size != name2->size ||
+        memcmp(&name1->name[0], &name2->name[0], name1->size) != 0)
+    {
+        LOG_ERROR("Names mismatch between NV_GetPublic and TR_GetName");
+        return 1;
```
Comment (tstruk) [inline test/integration/esys-tr-fromTpmPublic-key.int.c]: memleak

File: test/integration/esys-tr-getName-hierarchy.int.c [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    name1.size = offset;
+
+    r = Esys_TR_GetName(ectx, ESYS_TR_RH_OWNER, &name2);
+    goto_if_error(r, "TR get name", error);
+
+    if (name1.size != name2->size ||
+        memcmp(&name1.name[0], &name2->name[0], name1.size) != 0)
+    {
+        LOG_ERROR("Names mismatch between NV_GetPublic and TR_GetName");
+        return 1;
```
Comment (tstruk) [inline test/integration/esys-tr-getName-hierarchy.int.c]: memleak

**Comment by tstruk** [comment]
@AndreasFuchsSIT in general looks ok. I added few comments.

File: include/tss2/tss2_tpm2_types.h [hunk 23]
```diff
@@ -787,6 +787,8 @@ typedef uint32_t TPMA_MODES;
 
 /* Definition of BYTE TPMI_YES_NO Type */
 typedef BYTE TPMI_YES_NO;
+#define TPM2_NO  0 /* a value of 0 */
+#define TPM2_YES 1 /* a value of 1 */
```
[the remark this answers is not in the document]
Comment (AndreasFuchsSIT) [reply include/tss2/tss2_tpm2_types.h]: The TSS spec says that if a TPM-type starts with TPM_ the TSS-type starts with TPM2_ and if something does not start with TPM then TSS shall prefix it with TPM2_

File: test/integration/esys-audit.int.c (hunk 8 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-audit.int.c]: done

File: test/integration/esys-audit.int.c (hunk 9 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-audit.int.c]: done

File: test/integration/esys-certify-creation.int.c (hunk 10 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-certify-creation.int.c]: We want to return 1 and not r in these cases.
So I could add a new macro or have the manual if(){} thing here.
Just for those reasons I prefer the current. Agree ?

File: test/integration/esys-certify.int.c (hunk 11 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-certify.int.c]: We want to return 1 and not r in these cases.
So I could add a new macro or have the manual if(){} thing here.
Just for those reasons I prefer the current. Agree ?

File: test/integration/esys-commit.int.c (hunk 12 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-commit.int.c]: done

File: test/integration/esys-lock.int.c (hunk 13 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-lock.int.c]: Done and some reformatting of function invocations too

File: test/integration/esys-pp-commands.int.c (hunk 14 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-pp-commands.int.c]: done

File: test/integration/esys-tpm-tests.int.c (hunk 15 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tpm-tests.int.c]: done

File: test/integration/esys-tpm-tests.int.c (hunk 16 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tpm-tests.int.c]: done

File: test/integration/esys-policy-regression.int.c (hunk 17 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-policy-regression.int.c]: done

File: test/integration/esys-tr-fromTpmPublic-nv.int.c (hunk 18 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tr-fromTpmPublic-nv.int.c]: done

File: test/integration/esys-tr-fromTpmPublic-nv.int.c (hunk 19 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tr-fromTpmPublic-nv.int.c]: done, also for name2 here...

File: test/integration/esys-tr-fromTpmPublic-nv.int.c (hunk 20 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tr-fromTpmPublic-nv.int.c]: done also name2

File: test/integration/esys-tr-getName-hierarchy.int.c (hunk 22 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tr-getName-hierarchy.int.c]: done

File: test/integration/esys-tr-fromTpmPublic-key.int.c (hunk 21 above)
Comment (AndreasFuchsSIT) [reply test/integration/esys-tr-fromTpmPublic-key.int.c]: done

**Comment by AndreasFuchsSIT** [comment]
@tstruk all comments addressed;
- TPM2_YES/_NO should remain as is since that's spec.
- IMHO I prefer goto since we don't want to return r, but 1 or 0 (see above)

File: include/tss2/tss2_tpm2_types.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (tstruk) [reply include/tss2/tss2_tpm2_types.h]: If that's enforced by the spec then fine, but which TSS spec states that? Is it this one: https://trustedcomputinggroup.org/wp-content/uploads/TSS_Overview_Common_Structures_Version-0.9_Revision-03_Review_030918.pdf

I could not find anything enforcing this. I'm referring to 
https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-Library-Part-2_Structures-01.46_PubRev.pdf
Table 39 - Definition of (BYTE) TPMI_YES_NO Type.

File: test/integration/esys-certify-creation.int.c (hunk 10 above)
Comment (tstruk) [reply test/integration/esys-certify-creation.int.c]: Sure, that's fine.

File: test/integration/esys-certify.int.c (hunk 11 above)
Comment (tstruk) [reply test/integration/esys-certify.int.c]: agree

File: include/tss2/tss2_tpm2_types.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (tstruk) [reply include/tss2/tss2_tpm2_types.h]: @AndreasFuchsSIT how do we want to progress with this YES NO type.

[the remark this answers is not in the document]
Comment (tstruk) [reply include/tss2/tss2_tpm2_types.h]: That's fine. I'm not too religious about naming. My opinion is that TPM2_YES/NO it's an awful choice and people will use 1/0 instead anyway.

**Pull request merged**